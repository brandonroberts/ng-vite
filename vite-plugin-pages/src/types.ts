import type { Awaitable } from '@antfu/utils'

export type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>

export type ImportMode = 'sync' | 'async'
export type ImportModeResolver = (filepath: string, pluginOptions: ResolvedOptions) => ImportMode

export type CustomBlock = Record<string, any>

export type SupportedPagesResolver = 'vue' | 'react' | 'solid'

export interface PageOptions {
  dir: string
  baseRoute: string
}

/**
 * Plugin options.
 */
interface Options {
  /**
   * Paths to the directory to search for page components.
   * @default 'src/pages'
   */
  dirs: string | (string | PageOptions)[]
  /**
   * Valid file extensions for page components.
   * @default ['vue', 'js']
   */
  extensions: string[]
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[]
  /**
   * Import routes directly or as async components
   * @default 'root index file => "sync", others => "async"'
   */
  importMode: ImportMode | ImportModeResolver
  /**
   * Sync load top level index file
   * @default true
   * @deprecated use `importMode` instead
   */
  syncIndex: boolean
  /**
   * Use Nuxt.js style route naming
   * @default false
   * @deprecated use `routeStyle` instead
   */
  nuxtStyle: boolean
  /**
   * Use routing style
   * @default false
   */
  routeStyle: 'next' | 'nuxt' | 'remix'
  /**
   * Use case for route paths
   * @default false
     */
  caseSensitive: boolean
  /**
   * Set the default route block parser, or use `<route lang=xxx>` in SFC route block
   * @default 'json5'
   */
  routeBlockLang: 'json5' | 'json' | 'yaml' | 'yml'
  /**
   * Generate React Route
   * @default 'auto detect'
   */
  resolver: SupportedPagesResolver
  /**
   * Extend route records
   */
  extendRoute?: (route: any, parent: any | undefined) => any | void
  /**
   * Custom generated routes
   */
  onRoutesGenerated?: (routes: any[]) => Awaitable<any[] | void>
  /**
   * Custom generated client code
   */
  onClientGenerated?: (clientCode: string) => Awaitable<string | void>

  /**
   * Paths to the directory to search for page components.
   * @deprecated use `dirs` instead
   */
  pagesDir: string | (string | PageOptions)[]
  /**
   * Replace '[]' to '_' in bundle filename
   * @deprecated issue #122
   */
  replaceSquareBrackets: never
}

export type UserOptions = Partial<Options>

export interface ResolvedOptions extends Omit<Options, 'pagesDir' | 'replaceSquareBrackets' | 'nuxtStyle' | 'syncIndex'> {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string
  /**
   * Resolved page dirs
   */
  dirs: PageOptions[]
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp
}
