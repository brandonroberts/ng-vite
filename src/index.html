<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>NgVite</title>
  <base href="/">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="favicon.ico">
  <link rel="stylesheet" href="styles.module.css">
  <!-- <style>
    body {
      color: blue;
    }    
  </style> -->
</head>
<body>
  <app-root></app-root>
  <script type="module" src="main.ts"></script>
</body>
</html>
