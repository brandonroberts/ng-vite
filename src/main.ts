import { enableProdMode } from '@angular/core';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import '../node_modules/@angular/material/prebuilt-themes/deeppurple-amber.css';
import { AppModule } from './app/app.module';
import { environment } from './environments/environment';
import './styles.css';

if (environment.production) {
  enableProdMode();
}

platformBrowserDynamic().bootstrapModule(AppModule)
  .catch(err => console.error(err));
