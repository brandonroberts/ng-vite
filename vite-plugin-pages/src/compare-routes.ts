export function getPathSegments(route: any): string[] {
  const sanitizedPath = route.path.replace(/^\//, '').replace(/(?:\/$)/, '');
  return sanitizedPath ? sanitizedPath.split('/') : [];
}

/**
 * Compares two routes and returns sorting number
 * 0 - equal
 * -1 - `a` has priority over `b`
 * 1 - `b` has priority over `a`
 *
 * @param a Route
 * @param b Route
 */
 export const compareRoutes = (a: any, b: any): number => {
  const aSegments = getPathSegments(a);
  const bSegments = getPathSegments(b);

  for (let i = 0; i < Math.max(aSegments.length, bSegments.length); i++) {
    const current = compareSegments(aSegments, bSegments, i);
    if (current) {
      return current;
    }
  }
  // when paths are same, exact has priority
  return -1;
};

function compareSegments(
  aSegments: string[],
  bSegments: string[],
  index: number
): number {
  // if a has no more segments -> return 1
  if (aSegments.length <= index) {
    return 1;
  }
  // if b has no more segments -> return -1
  if (bSegments.length <= index) {
    return -1;
  }
  // if b is a catch-all segment
  if (aSegments[index].startsWith('[...') || bSegments[index].startsWith('[...')) {
    console.log(aSegments[index], bSegments[index]);
    return 1;
  }

  if (aSegments[index] === bSegments[index]) {
    return 0;
  }
  // prioritize non-empty path over empty
  if (!aSegments[index]) {
    return 1;
  }
  if (!bSegments[index]) {
    return -1;
  }
  // ignore param names
  if (isParam(aSegments[index]) && isParam(bSegments[index])) {
    return 0;
  }
  // static segment has priority over param
  if (isParam(aSegments[index])) {
    return 1;
  }
  if (isParam(bSegments[index])) {
    return -1;
  }
  // when all is same run string comparison
  return aSegments[index].localeCompare(bSegments[index]);
}

function isParam(segment: string): boolean {
  return segment.startsWith(':');
}