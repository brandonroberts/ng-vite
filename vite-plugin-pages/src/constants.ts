export const MODULE_IDS = [
  '~pages',
  '~react-pages',
  '~solid-pages',
  'pages-generated',
  'virtual:generated-pages',
  'virtual:generated-pages-react',
]

export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages'
export const ROUTE_BLOCK_ID_VIRTUAL = '/@vite-plugin-pages/route-block'

export const routeBlockQueryRE = /\?vue&type=route/

export const dynamicRouteRE = /^\[(.+)\]$/
export const cacheAllRouteRE = /^\[\.{3}(.*)\]$/
export const replaceDynamicRouteRE = /^\[(?:\.{3})?(.*)\]$/

export const nuxtDynamicRouteRE = /^_(.*)$/
export const nuxtCacheAllRouteRE = /^_$/

export const countSlashRE = /\//g

export const pathToNameRE = [
  /[_.\-\\/]/g,
  /[[:\]()]/g,
]

export const replaceIndexRE = /\/?index$/
