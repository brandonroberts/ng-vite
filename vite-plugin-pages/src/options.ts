import { resolve } from 'path'
import { slash, toArray } from '@antfu/utils'
import { getPageDirs } from './files'

import type { ImportModeResolver, ResolvedOptions, UserOptions } from './types'

function resolvePageDirs(dirs: UserOptions['dirs'], root: string, exclude: string[]) {
  dirs = toArray(dirs)
  return dirs.flatMap((dir) => {
    const option = typeof dir === 'string'
      ? { dir, baseRoute: '' }
      : dir

    option.dir = slash(resolve(root, option.dir)).replace(`${root}/`, '')
    option.baseRoute = option.baseRoute.replace(/^\//, '').replace(/\/$/, '')

    return getPageDirs(option, root, exclude)
  })
}

export const syncIndexResolver: ImportModeResolver = (filepath, options) => {
  for (const page of options.dirs) {
    if (page.baseRoute === '' && filepath.startsWith(`/${page.dir}/index`))
      return 'sync'
  }
  return 'async'
}

const getExtensions = (resolver: ResolvedOptions['resolver']) => {
  switch (resolver) {
  case 'vue':
    return ['vue', 'ts', 'js']
  case 'react':
  case 'solid':
    return ['tsx', 'jsx', 'ts', 'js']
  default:
    throw new Error(`Unsupported resolver: ${resolver}`)
  }
}

export function resolveOptions(userOptions: UserOptions, viteRoot?: string): ResolvedOptions {
  const {
    dirs = userOptions.pagesDir || ['src/pages'],
    routeBlockLang = 'json5',
    exclude = [],
    caseSensitive = false,
    resolver = 'vue',
    syncIndex = true,
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
  } = userOptions

  const root = viteRoot || slash(process.cwd())

  const importMode = userOptions.importMode || (syncIndex ? syncIndexResolver : 'async')

  const extensions = userOptions.extensions || getExtensions(resolver)

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`)

  const resolvedDirs = resolvePageDirs(dirs, root, exclude)

  const routeStyle = userOptions.nuxtStyle ? 'nuxt' : userOptions.routeStyle || 'next'

  const resolvedOptions: ResolvedOptions = {
    dirs: resolvedDirs,
    routeStyle,
    routeBlockLang,
    root,
    extensions,
    importMode,
    exclude,
    caseSensitive,
    resolver,
    extensionsRE,
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
  }

  return resolvedOptions
}
