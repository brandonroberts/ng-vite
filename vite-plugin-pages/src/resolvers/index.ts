export * from './vue'
export * from './react'
export * from './solid'
