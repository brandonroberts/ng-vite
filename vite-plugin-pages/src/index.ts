import { MODULE_IDS, MODULE_ID_VIRTUAL, ROUTE_BLOCK_ID_VIRTUAL, routeBlockQueryRE } from './constants'
import { PageContext } from './context'

import type { UserOptions } from './types'
import type { Plugin } from 'vite'

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let ctx: PageContext

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    async configResolved(config) {
      // auto set resolver for react project
      if (
        !userOptions.resolver
        && config.plugins.find(i => i.name.includes('vite:react'))
      )
        userOptions.resolver = 'react'

      // TODO: auto set resolver for solid project
      if (
        !userOptions.resolver
        && config.plugins.find(i => i.name.includes('solid'))
      )
        userOptions.resolver = 'solid'

      ctx = new PageContext(userOptions, config.root)
      ctx.setLogger(config.logger)
      await ctx.searchGlob()
    },
    configureServer(server) {
      ctx.setupViteServer(server)
    },
    resolveId(id) {
      if (MODULE_IDS.includes(id))
        return MODULE_ID_VIRTUAL

      if (routeBlockQueryRE.test(id))
        return ROUTE_BLOCK_ID_VIRTUAL

      return null
    },
    async load(id) {
      if (id === MODULE_ID_VIRTUAL)
        return ctx.resolveRoutes()

      if (id === ROUTE_BLOCK_ID_VIRTUAL) {
        return {
          code: 'export default {};',
          map: null,
        }
      }

      return null
    },
  }
}

export * from './types'
export {
  VueRoute,
  ReactRoute,
  SolidRoute,
} from './resolvers'
export { PageContext }
export default pagesPlugin
