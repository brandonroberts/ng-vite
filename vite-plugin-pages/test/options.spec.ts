import { describe, expect, test } from 'vitest'
import { resolveOptions } from '../src/options'

describe('Options', () => {
  test('resolve', () => {
    const options = resolveOptions({
      dirs: 'examples/vue/src/pages',
    })
    const reactOptions = resolveOptions({
      dirs: 'examples/react/src/pages',
      resolver: 'react',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    }, 'vue')
    expect(reactOptions).toMatchSnapshot({
      root: expect.any(String),
    }, 'react')
  })
})
