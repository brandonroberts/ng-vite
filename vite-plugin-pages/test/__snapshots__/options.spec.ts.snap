// Vitest Snapshot v1

exports[`Options resolve > react 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/react/src/pages",
    },
  ],
  "exclude": [],
  "extendRoute": undefined,
  "extensions": [
    "tsx",
    "jsx",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": "react",
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeStyle": "next",
}
`;

exports[`Options resolve > solid 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/solid/src/pages",
    },
  ],
  "exclude": [],
  "extendRoute": undefined,
  "extensions": [
    "tsx",
    "jsx",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": "solid",
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeStyle": "next",
}
`;

exports[`Options resolve > vue 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/vue/src/pages",
    },
  ],
  "exclude": [],
  "extendRoute": undefined,
  "extensions": [
    "vue",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(vue\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": "vue",
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeStyle": "next",
}
`;
